src/editor/editor.ml

```diff
@@ -0,0 +1,659 @@
+      ~gettimeofday
+      ~on_update
+      ~max_size: 99
+      ~snapshot () in
+
+  let toplevel_launch =
+    Learnocaml_toplevel.create
+      ~after_init ~timeout_prompt ~flood_prompt
+      ~on_disable_input: (fun _ -> disable_button_group toplevel_buttons_group)
+      ~on_enable_input: (fun _ -> enable_button_group toplevel_buttons_group)
+      ~container:(find_component "learnocaml-exo-toplevel-pane")
+      ~history () in
+  init_tabs () ;
+  toplevel_launch >>= fun top ->
+
+  (* ---- toplevel pane ------------------------------------------------- *)
+
+  begin toplevel_button
+      ~group: toplevel_buttons_group
+      ~icon: "cleanup" [%i"Clear"] @@ fun () ->
+    Learnocaml_toplevel.clear top ;
+    Lwt.return ()
+  end ;
+  begin toplevel_button
+      ~icon: "reload" [%i"Reset"] @@ fun () ->
+    (* toplevel_launch >>= fun top -> SHOULD BE UNNECESSARY *)
+     disabling_button_group toplevel_buttons_group
+         (fun () -> Learnocaml_toplevel.reset top)
+  end ;
+  begin toplevel_button
+      ~group: toplevel_buttons_group
+      ~icon: "run" [%i"Eval phrase"] @@ fun () ->
+    Learnocaml_toplevel.execute top ;
+    Lwt.return ()
+  end ;
+
+  (* ---- prelude pane --------------------------------------------------- *)
+
+  let editor_prelude = find_component "learnocaml-exo-prelude-pane" in
+  let editor_prel = Ocaml_mode.create_ocaml_editor
+                      (Tyxml_js.To_dom.of_div editor_prelude) in
+  let ace_prel = Ocaml_mode.get_editor editor_prel in
+  let contents= get_prelude id in
+  Ace.set_contents ace_prel contents ;
+  Ace.set_font_size ace_prel 18;
+
+  let typecheck_prelude () =
+    Editor_lib.typecheck true ace_prel editor_prel top "" (Ace.get_contents ace_prel) in
+  begin prelude_button
+      ~group: toplevel_buttons_group
+      ~icon: "typecheck" [%i"Check"] @@ typecheck_prelude
+  end;
+
+  (* ---- prepare pane --------------------------------------------------- *)
+
+  let editor_prepare = find_component "learnocaml-exo-prepare-pane" in
+  let editor_prep = Ocaml_mode.create_ocaml_editor
+                      (Tyxml_js.To_dom.of_div editor_prepare) in
+  let ace_prep = Ocaml_mode.get_editor editor_prep in
+  let contents= get_prepare id in
+  Ace.set_contents ace_prep contents ;
+  Ace.set_font_size ace_prep 18;
+
+  let typecheck_prepare () =
+    let prel = Ace.get_contents ace_prel ^ "\n" in
+    Editor_lib.typecheck true ace_prep editor_prep top prel
+      ~onpasterr:(fun () -> select_tab "prelude"; typecheck_prelude ())
+      (Ace.get_contents ace_prep) in
+  begin prepare_button
+      ~group: toplevel_buttons_group
+      ~icon: "typecheck" [%i"Check"] @@ typecheck_prepare
+  end;
+
+  (*-------question pane  -------------------------------------------------*)
+  let open Omd_representation in
+  let override_url = function
+    | Url(href,s,title) ->
+       Some ( let title_url = if title <> "" then
+                                String.concat " " [" title='";title;"'"]
```

Comment:
Nitpick: add spaces after commas and semicolons (and around binary operators);
and if I'm not mistaken, the `String.concat " "` should be `String.concat ""` here(?)

```suggestion
    | Url (href, s, title) ->
       Some (let title_url = if title <> "" then
                                String.concat "" [" title='"; title; "'"]
```